Load bboxes dataset from VIA tracks file (3/4)

sfmig opened the pull request:
Rebase after #234 

## Description

**What is this PR**

- [ ] Bug fix
- [x] Addition of a new feature
- [ ] Other

**Why is this PR needed?**
To be able to load a VIA tracks file with bounding boxes into a movement dataset.

**What does this PR do?**
- Adds a `movement.io.load_bboxes` module, which follows the equivalent poses one as much as possible.
- Adds corresponding unit tests in `tests_load_bboxes`.

**Question:** how to make `mypy` aware of the type transformations that take place in  `__attrs_post_init__`? 
For example, the confidence array passed to `ValidBboxesDataset` can be `None`, and `mypy` flags that a `.shape` attribute is used later which `None` doesn't have. But `mypy` seems to be missing that the confidence array is populated with nans in `__attrs_post_init__` if `None` is passed as input. Is there a nice way to fix this?

## References
This PR would close #167 

## How has this PR been tested?

Tests pass locally and on CI.

## Is this a breaking change?

No.

## Does this PR require an update to the documentation?

I updated `api_index.rst`.

## Checklist:

- [x] The code has been tested locally
- [x] Tests have been added to cover all new functionality
- [x] The documentation has been updated to reflect any changes
- [x] The code has been formatted with [pre-commit](https://pre-commit.com/)

VascoSch92 commented on movement/validators/datasets.py:
```diff
@@ -307,3 +308,12 @@ def __attrs_post_init__(self):
+                -1, 1
+            )
+            log_warning(
+                "Confidence array was not provided."
```
```suggestion
                "Confidence array was not provided. \n"
```
or? You wanted the comment to be printed on two lines?

sfmig replied:
It's fine in one line, but there was actually a space missing after the full stop. Fixed as part of [this PR](https://github.com/neuroinformatics-unit/movement/pull/230) - thanks!

VascoSch92 commented on movement/validators/files.py:
```diff
@@ -198,3 +201,210 @@ def csv_file_contains_expected_levels(self, attribute, value):
+        - Checking that there are as many unique frame numbers as unique image
+          files.
+
+        If the frame number is included as part of the image file name, it is
```
```suggestion
        If the frame number is included as part of the image file name, then it is
```

niksirbi commented on movement/io/load_bboxes.py:
```diff
@@ -0,0 +1,520 @@
+        boxes shapes, confidence scores and associated metadata.
+
+    """
+    # Create the time coordinate
+    time_coords = data.frame_array
+    time_unit = "frames"
+    # if fps is provided: time_coords is expressed in seconds.
+    if data.fps is not None:
+        # Compute frames from the start (first frame is frame 0).
+        # Ignoring type error because `data.frame_array` is not None after
+        # ValidBboxesDataset.__attrs_post_init__()
+        time_coords = np.arange(data.frame_array.shape[0], dtype=int)  # type: ignore
+        time_coords = time_coords / data.fps
+        time_unit = "seconds"
```
I'm confused about the need for and the nature of `frame_array`.
We may have discussed this but I forgot and would like to discuss it again.
Is it because the tracking data may not correspond to the entire source video, but instead to a subset of it? For example if the video is 1000 frames long, and only the 2nd half is tracked, then we would have `frame_array=np.array([500, 501, 502, ..., 999])`?

If my assumption is correct and the array consists of contiguous numbers, would it not just suffice to supply a `frame_offset` (i.e. a single integer value) and then calculate the length from the length of the `position` array? In that case the `ValidBboxesDataset` can be made to accept `frame_offset` instead of `frame_array`.

Moreover, there is a bit of an inconsistency in your current implementation, because frame numbers are relative to the video, while seconds are relative to the start of tracking. Namely, if a user supplied the `fps`, then time coordinates will start from `0` i.e. from the beginning of tracking, whereas is `fps` is missing they will start from the first tracked frame (`frame_array[0]` or `frame_offset`).

If my assumption is wrong and the `frame_array` can be non-contiguous (e.g. `[500, 505, 770]`), then the way you translate frames to time here also wouldn't make sense, because for this example array and `fps=40` we would get `time_coords=[0, 0.025, 0.05]`. Non-contiguous frames are useful for labels, so were you perhaps anticipating labelled or non-tracked data?

In general, `movement` currently can't really handle non-contiguous frames or non-constant inter-frame intervals (assumes fixed `fps`). For a proper handling of time and frames, we would first have to solve https://github.com/neuroinformatics-unit/movement/issues/117. I'm happy to brainstorm this together. I thinks that issue should be prioritised, because it's also essential for loading labelled data.

sfmig replied:
We talked about this offline but I thought it would be good to make a note here to keep a record of our conclusions.

> Is it because the tracking data may not correspond to the entire source video, but instead to a subset of it? For example if the video is 1000 frames long, and only the 2nd half is tracked, then we would have frame_array=np.array([500, 501, 502, ..., 999])?

Yes. Right now `frame_array` is optional, so if we don't pass it the frames will be numbered from 0 to `n`.

> would it not just suffice to supply a `frame_offset`?

Yes, that would be an alternative. In our discussion we weren't sure if one or other option was significantly advantageous. We thought having a `frame_array` may make it easier to adapt the code to load manually labelled frames, so we decided to stick with that approach for now.

> Moreover, there is a bit of an inconsistency in your current implementation, because frame numbers are relative to the video, while seconds are relative to the start of tracking

Yes, that is correct. I thought that was more intuitive, but it is also inconsistent. We agreed to make the time dimension to always have the same origin, whether it is expressed in frames or seconds. So if the video is only tracked from frame 500 to 999, and we would like to express time in seconds, we assume the origin of time (time = 0 seconds) is when frame 0 is captured.

> If my assumption is wrong and the frame_array can be non-contiguous...

We decided to add a validator check that ensures the frames passed via `frame_array` are contiguous

niksirbi commented on tests/test_unit/test_load_bboxes.py:
```diff
@@ -0,0 +1,307 @@
+    """Test that the function correctly extracts the confidence values from
+    the VIA dataframe.
+    """
+    df = request.getfixturevalue(df_input)
+
+    confidence_array = load_bboxes._extract_confidence_from_via_tracks_df(df)
+
+    assert np.array_equal(confidence_array, expected_array, equal_nan=True)
```
why the extra blank lines in this and the following test?

niksirbi reviewed (changes requested):
Thanks a lot @sfmig and well done on closely following the code architecture for loading poses. This makes me hopeful that we will be able to refactor them in future, such that we abstract away there common bits (to reduce repetition in validators and tests between poses and bboxes). But this shouldn't worry us right now.

I have two substantial comments apart form the cosmetic/trivial issues I've highlighted in specific comments.
- I don't fully understand what `frame_array` is needed for and how it is handled. See my specific comment about this. Perhaps we can discuss it in person together?
- The documentaiton pages for `input_output.md` and `movement_dataset.md` need to be updated. Specifically:
  - the new format has to be aded in the "Supported formats" section of Input/Output. 
  - A new "Loading tracks of bounding boxes" section has to be added to Input/Output.
  - The "Movement Dataset" page has to also include information about the bboxes dataset format, especially where it differs from the poses format

I'm happy for you to leave the documentation changes for a future PR (just open an issue in that case).

niksirbi commented on movement/io/load_bboxes.py:
```diff
@@ -0,0 +1,520 @@
+            "time_unit": time_unit,
+            "source_software": data.source_software,
+            "source_file": None,
+            "ds_type": "bboxes",
```
I think you should also modify the corresponding function for poses, to add a `"ds_type": "poses"` attribute.

sfmig commented:
From chats with @niksirbi:
We agreed to take the frame numbers as specified in the csv file for now (that is, not necessarily starting from 0), but to make the time coordinate have the same origin as the frame number.

sfmig commented on movement/io/load_bboxes.py:
```diff
@@ -0,0 +1,520 @@
+    Returns
+    -------
+    xarray.Dataset
+        ``movement`` bounding boxes dataset containing the boxes tracks,
```
After a quick slack survey I'm going for "tracks of the bounding boxes' centroids".

But let me know if you find that confusing and we can think about it a bit more

sfmig commented on movement/io/load_bboxes.py:
```diff
@@ -0,0 +1,520 @@
+    ds.attrs["source_software"] = "VIA-tracks"
+    ds.attrs["source_file"] = file.path.as_posix()
+
+    logger.info(f"Loaded bounding boxes' tracks from {via_file.path}:")
```
To be consistent with the choice above, I went for:
```python
logger.info(f"Loaded tracks of the bounding boxes from {via_file.path}:")
```

But again, let me know if you disagree or find it awkward.

sfmig commented:
Thanks @niksirbi for the feedback!

I think I addressed all the comments, let me know if something is missing.


#### `frame_array` and frame numbers
I did some changes to hopefully make more understandable the behaviour of keeping / resetting the frame numbers:
- In all the loaders of bboxes data (`from_file`, `from_via_tracks_file`, `from numpy`), the default behaviour is that the origin of the time dimension is the first loaded (aka tracked) frame, which would be frame number 0 captured at t = 0 seconds. This is to stay consistent with the pose data, and it is likely the most natural for our users.

- In the `from_file` and the `from_via_tracks_file` I added a `use_frame_numbers_from_file` option, that is `False` by default. If `True`, it will take the frame numbers from the input file, with whatever assumption for the time origin they have (they may assume for example the first frame is 0, or 1). 
    - The caveat is that if you take the frame number from the file, and later transform it to seconds, then the time origin is assumed to be at frame number 0 and time = 0 seconds. This is because we simply transform to seconds as `frame_number/fps`. 
    - Another option could be to add an extra attribute / argument to set the time origin manually, but I didn't want to complicate it too much. Let me know if it is clear as is, if not I am happy to give it another go.
    - If users want to keep track of the frame relative to the full video at which the analysed clip starts, they could assign a custom attribute. I sneakily added this in PR #245 - let me know thoughts.
    
- For the `from_numpy` loader, this is simply done via the `frame_array` input. If passed, those frame numbers are used, but the input is None by default, which means the frame numbers are assigned from 0 to N-1, with N being the first dimension of the position array.


#### Update Getting started documentation
- I moved this to a separate PR #245 

#### Extend tests to ensure postprocessing methods work with bboxes datasets
- Working on it in this (currently draft) PR #246 

How do you think we should merge these PRs? 
I have never done this but I'm wondering if we should merge PR #245 and #246 into this PR's branch, before merging this into `main`. Or should we use a `dev` branch?

niksirbi commented on movement/io/load_bboxes.py:
```diff
@@ -0,0 +1,644 @@
+    individual_names : list of str, optional
+        List of individual names for the tracked bounding boxes in the video.
+        If None (default), bounding boxes are assigned names based on the size
+        of the `position_array`. The names will be in the format of `id_<N>`,
+        where <N>  is an integer from 0 to `position_array.shape[1]-1` (i.e.,
```
reminder that if you want monospace formatting in rst you need double backticks.

niksirbi replied:
unless you deliberately wanted italics.

niksirbi commented on movement/validators/datasets.py:
```diff
@@ -197,21 +197,22 @@ class ValidBboxesDataset:
         List of individual names for the tracked bounding boxes in the video.
         If None (default), bounding boxes are assigned names based on the size
         of the `position_array`. The names will be in the format of `id_<N>`,
-        where <N>  is an integer from 1 to `position_array.shape[1]`.
+        where <N>  is an integer from 0 to `position_array.shape[1]-1`.
```
Reminder for double backticks if you want monospace formatting in rst.

niksirbi commented on movement/io/load_bboxes.py:
```diff
@@ -0,0 +1,644 @@
+    frame_array : np.ndarray, optional
+        Array of shape (n_frames, 1) containing the frame numbers for which
+        bounding boxes are defined. If None (default), frame numbers will
+        be assigned based on the first dimension of the `position_array`,
```
double backticks.

niksirbi commented on movement/io/load_bboxes.py:
```diff
@@ -0,0 +1,644 @@
+        the ``time`` coordinates are in seconds, they will indicate the
+        elapsed time from the capture of the first frame (assumed to be frame
+        0).
+    use_frame_numbers_from_file : bool, optional
+        If True, the frame numbers in the resulting dataset are
+        the same as the ones specified for each tracked bounding box in the
+        input file. This may be useful if the bounding boxes are tracked for a
+        subset of frames in a video, but you want to maintain the start of the
+        full video as the time origin. If False (default), the frame numbers
+        in the VIA tracks .csv file are instead mapped to a 0-based sequence of
+        consecutive integers.
```
Nice, extremely clear!

niksirbi commented on movement/io/load_bboxes.py:
```diff
@@ -0,0 +1,644 @@
+        ``movement`` dataset containing the position, shape, and confidence
+        scores of the tracked bounding boxes, and any associated metadata.
+
+    Examples
```
Thanks for all these examples, they're very helpful and unambiguous.

niksirbi commented on movement/io/load_bboxes.py:
```diff
@@ -0,0 +1,644 @@
+
+    >>> from movement.io import load_bboxes
+    >>> ds = load_bboxes.from_file(
+    ...     "path/to/file.csv",
+    ...     source_software="VIA-tracks",
+    ... )
+
+    Create a dataset from the VIA tracks .csv file at "path/to/file.csv", with
+    the time coordinates in seconds, and assuming t = 0 seconds corresponds to
+    the first tracked frame in the file.
+
+    >>> from movement.io import load_bboxes
+    >>> ds = load_bboxes.from_file(
+    ...     "path/to/file.csv",
+    ...     source_software="VIA-tracks",
+    ...     fps=30,
+    ... )
+
+    Create a dataset from the VIA tracks .csv file at "path/to/file.csv", with
+    the time coordinates in frames, and using the same frame numbers as
+    in the VIA tracks .csv file.
+
+    >>> from movement.io import load_bboxes
+    >>> ds = load_bboxes.from_file(
+    ...     "path/to/file.csv",
+    ...     source_software="VIA-tracks",
+    ...     use_frame_numbers_from_file=True.
+    ... )
+
+    Create a dataset from the VIA tracks .csv file at "path/to/file.csv", with
+    the time coordinates in seconds, and assuming t = 0 seconds corresponds to
+    the 0th frame in the full video.
+
+    >>> from movement.io import load_bboxes
+    >>> ds = load_bboxes.from_file(
+    ...     "path/to/file.csv",
+    ...     source_software="VIA-tracks",
+    ...     fps=30,
+    ...     use_frame_numbers_from_file=True,
+    ... )
```
Because most of these examples are specific to the VIA-tracks format, maybe it's better to move them to the docstring of the `from_via_tracks_file`? Instead you can use a generic simple example here (basically swap the examples between the two functions). 

Since you alread have a "See also" section redirecting people to the `from_via_tracks_file` docstring, we can point people there for detailed format-specific examples/instructions.

Since this is currently the only bbox format we support, I don't feel very strongly about it, but may become relevant in the future.

sfmig replied:
yes, nice idea ✨

niksirbi reviewed (approved):
Thanks @sfmig, I'm happy with how you've handled things, and the updated docstring make it super clear what happens to time and frames.

I've only left very few suggestion re docstrings, feel free to adopt or not.

niksirbi commented:
> How do you think we should merge these PRs?
> I have never done this but I'm wondering if we should merge PR #245 and #246 into this PR's branch, before merging this into `main`. Or should we use a `dev` branch?

I've never done this as well. Maybe we should consult with Alessandro or Joe who have more experience with complex merges.
How about merging each of these 3 PRs separately (and sequentially) to main (with rebases in-between)? Is that not an option? It would be nice to have them as 3 separate commits in `main`, but if that' painful, feel free to merge any way that's convenient.
